migration/migration1_joblogger.go

```diff
@@ -0,0 +1,84 @@
+package migration
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	"github.com/textileio/powergate/ffs"
+	"github.com/textileio/powergate/util"
+)
+
+func migrateJobLogger(ds datastoreReaderWriter, cidOwners map[cid.Cid][]ffs.APIID) error {
+	q := query.Query{Prefix: "/ffs/joblogger"}
+	res, err := ds.Query(q)
+	if err != nil {
+		return fmt.Errorf("querying joblogger: %s", err)
+	}
+	defer func() { _ = res.Close() }()
+
+	var lock sync.Mutex
+	var errors []string
+	lim := make(chan struct{}, 1000)
+	for r := range res.Next() {
+		if r.Error != nil {
+			return fmt.Errorf("iterating result: %s", r.Error)
+		}
+		lim <- struct{}{}
+
+		r := r
+		go func() {
+			defer func() { <-lim }()
+
+			originalKey := datastore.NewKey(r.Key)
+			cidStr := originalKey.Namespaces()[2] // /ffs/joblogger/<cid>/<timestamp>
+			cid, err := util.CidFromString(cidStr)
+			if err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("parsing cid %s: %s", cidStr, err))
+				lock.Unlock()
+				return
+			}
+			timestampStr := originalKey.Namespaces()[3]
+
+			owners := cidOwners[cid]
+
+			// Step 1/2:
+			// For each cid owner, we create the same registry
+			// in the new key version.
+			for _, iid := range owners {
+				newKey := datastore.NewKey("/ffs/joblogger_v2/").ChildString(iid.String()).ChildString(cidStr).ChildString(timestampStr)
```

Comment:
Here the detail about `_v2` suffixes.
What's happening in many migrations is that we try doing something like:
- Original key: `/hello/world/<cid>`
- New key: `/hello/world/<iid>/<cid>`

For doing the migration, we scan all original keys with `Query{Prefix:"/hello/world"}`.

Depending on the underlying way that the `go-datastore` implementation works, that can produce problems; and that was the case, since the new keys we're `.Put`ing are also a result of the query itself.

That's to say, we start getting `/hello/world/<cid>` keys, but at the same time we're iterating that query we're inserting others `/hello/world/<iid>/<cid>` which will appear later in the query iteration and is *not* something we want (new keys satisfy the prefix from the query). We only want to iterate the original keys.

Not a big deal and is solved with the `_v2` suffixing to avoid the iteration to consider the new keys.

In the Badger implementation of `go-datastore` this doesn't happen, since Badger has snapshot capabilities on the query execution and you won't see new keys that appear after the query was started. But in Mongo that isn't the case. It's just different ways that queries work in different databases.

Reply:
Tricky